### DISCUSSION TITLE
What kind of simple "Hello World" type GUIs work ?

### DISCUSSION THREAD

**Discussion by constantitus** [darlinghq/darling#1550] [Q&A]
It is stated in the wiki that `only simple "Hello World" type GUIs will work.`. So I tried building a simple one with glfw.
Here's the boilerplate code
<details>
  <summary>test.c</summary>

```c
#include <stdio.h>
#include <stdlib.h>
#include <GLFW/glfw3.h>

static void error_callback(int error, const char *description) {
    fputs(description, stderr);
}

int main(void) {
    GLFWwindow *window;

    glfwSetErrorCallback(error_callback);

    if (!glfwInit())
        exit(EXIT_FAILURE);

    window = glfwCreateWindow(640, 480, "glfw boilerplate", NULL, NULL);
    if (!window) {
        glfwTerminate();
        exit(EXIT_FAILURE);
    }

    glfwMakeContextCurrent(window);

    while (!glfwWindowShouldClose(window)) {
        glfwSwapBuffers(window);
        glfwPollEvents();
        //glfwWaitEvents();
    }

    glfwDestroyWindow(window);

    glfwTerminate();
    exit(EXIT_SUCCESS);
}
```
</details>

It just tries to open a window, not even asking for an opengl context or anything.
Here's the build command aswell
<details>
  <summary>build command</summary>

```sh
clang \
    -o test test.c \
    -mmacosx-version-min=10.13 \
    --sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX10.13.sdk \
    -arch x86_64 \
    -framework Cocoa \
    -framework Opengl \
    -framework IOKit \
    -framework CoreVideo \
    -I./glfw/include \
    -L./glfw/lib-x86_64 \
    -lglfw3
```
</details>

It compiles, but running it gives me the following error
```
2024-10-16 02:53:30.063 test[1172:3e8] -[NSEvent addLocalMonitorForEventsMatchingMask:handler:] unimplemented in /home/runner/work/darling/darling/src/external/cocotron/AppKit/NSEvent.subproj/NSEvent.m at 355
dyld: lazy symbol binding failed: Symbol not found: _CGEventSourceSetLocalEventsSuppressionInterval
  Referenced from: /Volumes/SystemRoot/home/bern/playground/macos/test
  Expected in: /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics

dyld: Symbol not found: _CGEventSourceSetLocalEventsSuppressionInterval
  Referenced from: /Volumes/SystemRoot/home/bern/playground/macos/test
  Expected in: /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics

abort_with_payload: reason: Symbol not found: _CGEventSourceSetLocalEventsSuppressionInterval
  Referenced from: /Volumes/SystemRoot/home/bern/playground/macos/test
  Expected in: /System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics
; code: 4
```

I was just curious, is this supposed to work in darling ?
Unfortunately, I don't have a macos running system atm, so I can't test whether or not what I did is right or the compiled binary works at all on a mac, but I'm glad I was able to get it to at least compile.

**Comment by CuriousTommy** [comment]
> I was just curious, is this supposed to work in darling ?

We haven't really tested against 3rd party GUI frameworks. The best way to run a simple GUI on Darling is to directly use AppKit (just make sure you are not building a Storyboard app, those are not supported for now)

**Reply by p0358** [reply]
@CuriousTommy Actually speaking of 3rd party GUI frameworks, have you considered trying an approach that box64 or Hangover are doing? In their case it's for ARM emulation, but basically what they do is to wrap emulated calls to graphics libraries and execute them on the native counterparts on the host (whereas Hangover additionally also runs WINE on host). They do it for performance reasons, but I imagine that perhaps similar approach could be tried for Darling, but for the purpose of drawing GUI apps on Linux versions of the libraries, by forwarding calls from the Mac version? Might be long shot, since there's things like calling conventions in play, and that sometimes the APIs might slightly differ between the platforms etc. but perhaps it could be a reasonable starting point to have some more involved GUI apps running?
